src/chainstate/stacks/index/storage.rs

```diff
@@ -1363,122 +2159,191 @@ impl<'a, T: MarfTrieId> TrieStorageConnection<'a, T> {
         }
     }
 
+    /// Is the given block represented in either the confirmed or unconfirmed block tables?
+    /// The mined table is ignored.
     pub fn has_block(&self, bhh: &T) -> Result<bool, Error> {
         Ok(self.has_confirmed_block(bhh)? || self.has_unconfirmed_block(bhh)?)
     }
 
-    // used for providing a option<block identifier> when re-opening a block --
-    //   because the previously open block may have been the last_extended block,
-    //   id may have been None.
+    /// Used for providing a option<block identifier> when re-opening a block --
+    ///   because the previously open block may have been the uncommitted_writes block,
+    ///   id may have been None.
     pub fn open_block_maybe_id(&mut self, bhh: &T, id: Option<u32>) -> Result<(), Error> {
         match id {
             Some(id) => self.open_block_known_id(bhh, id),
             None => self.open_block(bhh),
         }
     }
 
-    // used for providing a block identifier when opening a block -- usually used
-    //   when following a backptr, which stores the block identifier directly.
+    /// Used for providing a block identifier when opening a block -- usually used
+    ///   when following a backptr, which stores the block identifier directly.
     pub fn open_block_known_id(&mut self, bhh: &T, id: u32) -> Result<(), Error> {
+        trace!(
+            "open_block_known_id({},{}) (unconfirmed={:?},{})",
+            bhh,
+            id,
+            &self.unconfirmed_block_id,
+            self.unconfirmed()
+        );
         if *bhh == self.data.cur_block && self.data.cur_block_id.is_some() {
             // no-op
             return Ok(());
         }
 
-        if let Some((ref last_extended, _)) = self.data.last_extended {
-            if last_extended == bhh {
+        if let Some((ref uncommitted_bhh, _)) = self.data.uncommitted_writes {
+            if uncommitted_bhh == bhh {
                 panic!("BUG: passed id of a currently building block");
             }
         }
 
         // opening a different Trie than the one we're extending
-        self.data.cur_block_id = Some(id);
-        self.data.cur_block = bhh.clone();
-
+        self.data.set_block(bhh.clone(), Some(id));
         Ok(())
     }
 
+    /// Open a trie's block, identified by `bhh`.  Updates the internal state to point to it, so
+    /// that all node reads will occur relative to it.
     pub fn open_block(&mut self, bhh: &T) -> Result<(), Error> {
+        trace!(
+            "open_block({}) (unconfirmed={:?},{})",
+            bhh,
+            &self.unconfirmed_block_id,
+            self.unconfirmed()
+        );
+        self.bench.open_block_start();
+
         if *bhh == self.data.cur_block && self.data.cur_block_id.is_some() {
             // no-op
+            if self.unconfirmed() {
+                if self.data.cur_block_id
+                    == trie_sql::get_unconfirmed_block_identifier(&self.db, bhh)?
+                {
```

Comment:
What does it mean if this equality is false?

Reply:
It means that we're opening a block that doesn't represent unconfirmed state.  This is normal over the course of a MARF walk from an unconfirmed trie.  A MARF walk would start out from the unconfirmed trie, encounter a backptr to a confirmed trie, and then open the confirmed trie.  In that last step of resolving the backptr, this condition would be `false`.

Note that it's impossible by construction to step _to_ an unconfirmed trie, because (1) the unconfirmed trie's block ID is globally unique and derived from the confirmed trie it extends, and (2) you can't build an unconfirmed trie on top of an unconfirmed trie.

Reply:
Recall that `self.unconfirmed()` means "this storage connection was instantiated for reading and writing unconfirmed state on top of a MARF."  This is important because unconfirmed trie blobs are treated differently -- for example, they cannot be stored to the new `TrieFile` instance, since they must later be deleted (and `TrieFile` only supports append-only writes).

Reply:
> Recall that self.unconfirmed() means "this storage connection was instantiated for reading and writing unconfirmed state on top of a MARF."

Can you add that to a rustdoc for the field `unconfirmed` and for the function `pub fn unconfirmed(&self)`?

Reply:
Sure!